azext_iot/iothub/providers/device_identity.py

```diff
@@ -0,0 +1,409 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from pathlib import PurePath
+from os import getcwd, makedirs
+from os.path import exists
+from shutil import rmtree
+from azext_iot.common.certops import (
+    create_self_signed_certificate,
+    create_signed_cert,
+    make_cert_chain,
+)
+
+from azext_iot.common.fileops import tar_directory, write_content_to_file
+from azext_iot.iothub.edge_device_config import (
+    DEVICE_README,
+    EDGE_DEVICE_BUNDLE_DEFAULT_FOLDER_NAME,
+    EDGE_ROOT_CERTIFICATE_FILENAME,
+    create_edge_device_config,
+    process_nested_edge_config_args,
+    process_nested_edge_config_file_content,
+    create_nested_edge_device_config_script,
+)
+from tqdm import tqdm
+from time import sleep
+from typing import Dict, List
+from knack.log import get_logger
+from typing import Optional
+from azext_iot._factory import SdkResolver
+from azext_iot.common.shared import (
+    DeviceAuthType,
+    SdkType,
+    NestedEdgeConfig,
+)
+from azext_iot.iothub.providers.base import IoTHubProvider
+from azext_iot.common.utility import (
+    process_json_arg,
+    process_yaml_arg,
+)
+from azext_iot.operations.generic import _execute_query
+from azure.cli.core.azclierror import (
+    AzureResponseError,
+    InvalidArgumentValueError,
+    MutuallyExclusiveArgumentError,
+)
+from azext_iot.operations.hub import _assemble_device
+from azext_iot.sdk.iothub.service.models import Device
+
+logger = get_logger(__name__)
+
+
+class DeviceIdentityProvider(IoTHubProvider):
+    def __init__(
+        self,
+        cmd,
+        hub_name: Optional[str] = None,
+        rg: Optional[str] = None,
+        login: Optional[str] = None,
+        auth_type_dataplane: Optional[str] = None,
+    ):
+        super(DeviceIdentityProvider, self).__init__(
+            cmd=cmd,
+            hub_name=hub_name,
+            rg=rg,
+            login=login,
+            auth_type_dataplane=auth_type_dataplane,
+        )
+
+        self.hub_resolver = SdkResolver(target=self.target)
+        self.service_sdk = self.hub_resolver.get_sdk(SdkType.service_sdk)
+
+    def create_edge_hierarchy(
+        self,
+        devices: Optional[List[List[str]]] = None,
+        config_file: Optional[str] = None,
+        visualize: Optional[bool] = False,
```

Comment:
nit, because the default value False is assigned, I don't see the semantic value that Optional provides for the bool.